### ISSUE TITLE
SCNd64 issue with CRAN for Windows

### ISSUE THREAD

**Issue by gillian-iris** [iris-edu/libmseed#79]
Requested changes to avoid CRAN check compilation warnings when submitting package that uses libmseed. The use of SCNd64 generates these warnings on their Windows builder:
warning: ISO C does not support the 'I64' ms_scanf length modifier [-Wformat=]

fileutils.c, line 471
      /*(replacing) sscanf (hdrstr, " %" SCNd64, &datasize);*/
      datasize = scan_d64(hdrstr,0,&endptr);

genutils.c, line 1220
      /*(replacing)fields  = sscanf (readline, "#@ %" SCNd64, &expires); */
      fields = 0;
      expires = scan_d64(readline,2,&endptr);
      if (endptr - readline + 2 > 0) fields++;

genutils.c, line 1244
    /*(replacing) fields = sscanf (readline, "%" SCNd64 " %d ", &leapsecond, &TAIdelta); */
    fields = 0;
    leapsecond = scan_d64(readline,0,&endptr);
    if (endptr - readline > 0) fields++;
    fields += sscanf(endptr, " %d ", &TAIdelta);

libmseed.h, inserted after line 811
extern int64_t scan_d64(char *str, int offset, char **endptr);

**Comment by chad-iris** [comment]
Thanks.  I assume the scan_d64 routine was added to the code somewhere.
 
Which compilers on Windows does the I64 this come from?  It'd be nice to have the options used as well, e.g. was -pedantic used?

**Comment by chad-iris** [comment]
Looks like they use MinGW-W64 on Windows.

**Comment by gillian-iris** [comment]
Here is the scan_d64 routine:

genutils.c, inserted after line 24
int64_t scan_d64(char *str, int offset, char **endptr);

genutils.c,  at end
/******************************************
 * scan_d64 -- scan string for 64-bit int
 * read string starting at character offset
 *     and convert to a 64 bit long long int.
 * return the long long int.
 * set endptr to where we stopped the scan.
 * REC - 2016-12-02
 *******************************************/
int64_t scan_d64(char *str, int offset, char **endptr) {

    /* implementation based on die.net strtol(3) man page example */
    int base = 10;
    int64_t val = 0;

    /* conversion to unsigned long long */
    errno = 0;
    val = strtoull(str+offset, endptr, base);

    if (errno != 0 && val == 0) {
        perror("strtoull");
        exit(EXIT_FAILURE);
    }

    return val;
}

**Comment by gillian-iris** [comment]
CRAN does not allow the use of -Wno-pedantic-ms-format because of portability issues.

**Comment by chad-iris** [comment]
> d:/Compiler/rtools40/mingw32/bin/gcc -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c fileutils.c -o fileutils.o

Hmmm.  I compiled on Windows with MSYS2 and MinGW64 with those options and got no warnings at all, output below.  Also, CRAN doesn't seem to have any mingw32 targets anymore. 

**Perhaps this modification is no longer needed?**

```
Chad@DESKTOP-6LQFLF8 MINGW64 /c/Users/chad/Downloads/libmseed-2.19.7
$ CFLAGS="-pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign" make
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c fileutils.c -o fileutils.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c genutils.c -o genutils.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c gswap.c -o gswap.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c lmplatform.c -o lmplatform.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c lookup.c -o lookup.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c msrutils.c -o msrutils.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c pack.c -o pack.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c packdata.c -o packdata.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c traceutils.c -o traceutils.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c tracelist.c -o tracelist.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c parseutils.c -o parseutils.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c unpack.c -o unpack.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c unpackdata.c -o unpackdata.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c selection.c -o selection.o
cc  -pedantic -O2 -Wall -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -c logging.c -o logging.o
Building static library libmseed.a
rm -f -f libmseed.a
ar -crs libmseed.a fileutils.o genutils.o gswap.o lmplatform.o lookup.o msrutils.o pack.o packdata.o traceutils.o tracelist.o parseutils.o unpack.o unpackdata.o selection.o logging.o

Chad@DESKTOP-6LQFLF8 MINGW64 /c/Users/chad/Downloads/libmseed-2.19.7
$ cc -v
Using built-in specs.
COLLECT_GCC=C:\msys64\mingw64\bin\cc.exe
COLLECT_LTO_WRAPPER=C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-11.2.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/include --libexecdir=/mingw64/lib --enable-bootstrap --enable-checking=release --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++,jit --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts --enable-libstdcxx-time --disable-libstdcxx-pch --disable-libstdcxx-debug --enable-lto --enable-libgomp --disable-multilib --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev10, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld --with-boot-ldflags='-pipe -Wl,--disable-dynamicbase -static-libstdc++ -static-libgcc' LDFLAGS_FOR_TARGET=-pipe --enable-linker-plugin-flags='LDFLAGS=-static-libstdc++\ -static-libgcc\ -pipe\ -Wl,--stack,12582912'
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Rev10, Built by MSYS2 project)

Chad@DESKTOP-6LQFLF8 MINGW64 /c/Users/chad/Downloads/libmseed-2.19.7
```

**Comment by gillian-iris** [comment]
I tested that earlier today, restoring SCNd64 to fileutils.c. It failed: https://win-builder.r-project.org/6GD74T73i3xE/00install.out

**Comment by chad-iris** [comment]
Weird that there is a 32-bit build in there and old compilers being used (gcc 4.9.3) given that the [check results are all 64-bit and newer](https://cran.r-project.org/web/checks/check_results_IRISSeismic.html) and the generic "[Package Check Flavors](https://cran.r-project.org/web/checks/check_flavors.html)" are also all 64-bit and newer.

After a lot of thrashing I found a way to install that old compiler and was able to duplicate the issue.  What I found is that it is the compiler/toolchain _itself_ injecting the `I64` because that's what's needed on older Windows.  There is no way for libmseed to override that to use the appropriate `SCNd64`, which _is_ part of the C99 standard being enforced with the 
`std=gnu99`.  This combination of platform, legacy toolchain, and pedantic option is tripping on itself.

I think this problem will go away when CRAN builds with newer compiler.  I don't think there is a better fix that what you've got now, which shouldn't go upstream.
